### DISCUSSION TITLE
Ignoring gclid and utm tags

### DISCUSSION THREAD

**Discussion by mthorsdal** [spatie/laravel-responsecache#382] [Q&A] [answered]
Hi Guys

I'm a happy user of laravel-responsecache, but I've noticed that all my visitors from AdWords rarely see the cached versions of my site - likely due to the appended ?gclid=XXXXX or ?utm_source=XXXX URL parameters.

Since these URL parameters do not have any impact on how the page should be displayed (in my case) nor need to be sent to any laravel controller, I'm wondering if it would be possible to ignore these tags and show the cached version of my page saved 

To be clear, what I want to achieve is this:

I have a page https://example.com/page which is cached using laravel-responsecache and I'd like to display the cached page when a user visits https://example.com/page?gclid=*

I'm wondering if anyone has had luck with this or could point me in the right direction?

Thanks!

**Comment by caocanhlinh** [comment]
I think you can 301 redirect to the original url after they click on the ad, and that way the cache will be used.

**Reply by mthorsdal** [reply]
Unfortunately that doesn't work as redirects to remove the gclid or UTM params cause the tracking on Google analytics to fail

**Comment by twf-nikhila** [comment]
@mthorsdal did you found any solution for this?

**Reply by mthorsdal** [reply]
Unfortunately not 😢

**Comment by vinhkhangphung** [comment, marked as answer]
I think you should take a look at `DefaultHasher` class and implement a custom Hasher to serve your purpose. The issue lies in the fact that the DefaultHasher by the package use the whole query string as part of the hash. You should do some stuff to remove out all unnecessary query string (gclid= or utm_source=). Also, remember to change the Hasher in the config file. I will try it out myself if I had the time.

DefaultHasher
```
protected function getNormalizedRequestUri(Request $request): string
{
    if ($queryString = $request->getQueryString()) {
        $queryString = '?'.$queryString;
    }

    return $request->getBaseUrl().$request->getPathInfo().$queryString;
}
```

Register hasher in config file
`'hasher' => \Spatie\ResponseCache\Hasher\DefaultHasher::class`

**Reply by RChutchev** [reply]
Yep, just use return without $queryString;

```php
protected function getNormalizedRequestUri(Request $request): string
    {
        return $request->getBaseUrl().$request->getPathInfo();
    }
```

As a result, caching ignores all parameters.

**Comment by mthorsdal** [comment]
That did the job, thanks! 

In my case, I wanted to exclude tracking parameters, so I used

```
$params = array_filter($params, function($key) {
                return strpos($key, 'utm_') !== 0 && $key !== 'gclid';
            }, ARRAY_FILTER_USE_KEY);
            
            if (!empty($params)) {
                $url .= '?' . http_build_query($params);
            }
```
